src/credentials/sso/ssoAccessTokenProvider.ts

```diff
@@ -0,0 +1,148 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as vscode from 'vscode'
+import { ClientRegistration } from './clientRegistration'
+import { AccessToken } from './accessToken'
+import { DiskCache } from './diskCache'
+import { Authorization } from './authorization'
+import { getLogger } from '../../shared/logger'
+
+const CLIENT_REGISTRATION_TYPE = 'public'
+const GRANT_TYPE = 'urn:ietf:params:oauth:grant-type:device_code'
+
+export class SsoAccessTokenProvider {
+    private ssoRegion: string
+    private ssoUrl: string
+    private ssoOidcClient: AWS.SSOOIDC
+    private cache: DiskCache
+
+    constructor(ssoRegion: string, ssoUrl: string, ssoOidcClient: AWS.SSOOIDC, cache: DiskCache) {
+        this.ssoRegion = ssoRegion
+        this.ssoOidcClient = ssoOidcClient
+        this.cache = cache
+        this.ssoUrl = ssoUrl
+    }
+
+    public async accessToken(): Promise<AccessToken> {
+        const accessToken = this.cache.loadAccessToken(this.ssoUrl)
+        if (accessToken) {
+            return accessToken
+        }
+
+        try {
+            const token = await this.pollForToken()
+            this.cache.saveAccessToken(this.ssoUrl, token)
+            return token
+        } catch (err) {
+            throw err
+        }
+    }
+
+    private async pollForToken(): Promise<AccessToken> {
+        const registration = await this.registerClient()
+        const authorization = await this.authorizeClient(registration)
+        const deviceCodeExpiration = Date.now() + authorization.expiresIn * 1000
+
+        //temporary logging to get URL
+        const logger = getLogger()
```

Comment:
The logger isn't reused so you could: `getLogger().info...`, but probably should add more logging statements below